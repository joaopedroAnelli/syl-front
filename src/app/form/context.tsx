'use client';

import { createContext, useState, useRef, FC, PropsWithChildren } from 'react';
import { FormContext as FormContextType, SubmitStatus } from './types';
import { useRouter } from 'next/navigation';
import * as Toast from '@radix-ui/react-toast';
import { XMarkIcon } from '@heroicons/react/24/solid';

const INITIAL_DATA = {
  address: '',
  city: '',
  state: '',
  offerPrice: 0,
  size: 0,
  sizeUnit: 'sqft',
  contact: {
    name: '',
    email: '',
  },
};
export const FormContext = createContext<FormContextType>({
  data: INITIAL_DATA,
  setData: () => {},
  pageFormRef: { current: null },
  submitStatus: SubmitStatus.IDLE,
  submitForm: () => Promise.resolve(true),
});

export type FormProviderProps = {};

export const FormProvider: FC<PropsWithChildren<FormProviderProps>> = ({
  children,
}) => {
  const [data, setData] = useState(INITIAL_DATA);
  const pageFormRef = useRef(null);
  const [submitStatus, setSubmitStatus] = useState<SubmitStatus>(
    SubmitStatus.IDLE
  );
  const [errorMessage, setErrorMessage] = useState<string | null>(null);

  const [toastOpen, setToastOpen] = useState(false);

  const router = useRouter();

  const submitForm = async () => {
    setSubmitStatus(SubmitStatus.LOADING);

    const cmsResponse = await fetch('/api/properties', {
      method: 'POST',
      body: JSON.stringify(data),
    });

    if (!cmsResponse.ok) {
      setSubmitStatus(SubmitStatus.ERROR);

      let message = await cmsResponse.text();

      if (!message) {
        message = cmsResponse.statusText;
      }

      setErrorMessage(message);

      setToastOpen(true);
      return false;
    }

    setSubmitStatus(SubmitStatus.SUCCESS);

    router.push('/form/success');

    return true;
  };

  return (
    <FormContext.Provider
      value={{
        data,
        setData,
        pageFormRef,
        submitStatus,
        submitForm,
      }}
    >
      {children}

      <Toast.Provider swipeDirection='right'>
        <Toast.Root
          open={toastOpen}
          onOpenChange={setToastOpen}
          className='ToastRoot bg-red-50 rounded-md p-4 shadow-md flex gap-3'
        >
          <div>
            <Toast.Title>
              <h3 className='text-red-700 font-bold text-lg'>Internal Error</h3>
            </Toast.Title>

            <Toast.Description className='text-red-900'>
              {errorMessage}
            </Toast.Description>
          </div>

          <Toast.Close aria-label='Close'>
            <span aria-hidden>
              <XMarkIcon className='w-5 h-5 absolute top-6 right-6 text-red-900' />
            </span>
          </Toast.Close>
        </Toast.Root>
        <Toast.Viewport className='fixed bottom-0 right-0 z-50 flex flex-col p-3' />
      </Toast.Provider>
    </FormContext.Provider>
  );
};
