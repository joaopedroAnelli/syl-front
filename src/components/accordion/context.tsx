'use client';

import { createContext, FC, PropsWithChildren, useState } from 'react';

export type AccordionContextType = {
  open: boolean;
  setOpen: (open: boolean) => void;
};

export const AccordionContext = createContext<AccordionContextType>({
  open: false,
  setOpen: () => {},
});

export const AccordionProvider: FC<PropsWithChildren<{}>> = ({ children }) => {
  const [open, setOpen] = useState(false);

  const changeOpen = (open: boolean) => {
    setOpen(open);
  };

  return (
    <AccordionContext.Provider value={{ open, setOpen: changeOpen }}>
      {children}
    </AccordionContext.Provider>
  );
};
