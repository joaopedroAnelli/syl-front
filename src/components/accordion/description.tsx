import { PropsWithChildren, FC } from 'react';

export type DescriptionProps = PropsWithChildren<{}>;

export const Description: FC<DescriptionProps> = ({ children }) => {
  return <p className='mx-4 text-off-white'>{children}</p>;
};
