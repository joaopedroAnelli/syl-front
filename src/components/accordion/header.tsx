import { FC, PropsWithChildren } from 'react';

export type HeaderProps = PropsWithChildren<{}>;

export const Header: FC<HeaderProps> = ({ children }) => {
  return <div className='flex items-center justify-between'>{children}</div>;
};
