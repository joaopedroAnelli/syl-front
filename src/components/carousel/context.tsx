'use client';
import {
  FC,
  PropsWithChildren,
  createContext,
  useContext,
  useEffect,
  useRef,
  useState,
} from 'react';

export type CarouselContext = {
  autoPlay: boolean;
  setAutoPlay: (autoPlay: boolean) => void;
  active: {
    index: number;
    id: string | number;
  };
  showNext: () => void;
  showPrev: () => void;
  registerItem: (id: string | number) => void;
};

export const CarouselContext = createContext<CarouselContext>({
  autoPlay: false,
  setAutoPlay: () => {},
  active: { index: 0, id: 'first' },
  showNext: () => {},
  showPrev: () => {},
  registerItem: () => {},
});

export type CarouselContextProviderProps = {
  autoPlay?: boolean;
  autoPlayDuration?: number;
};

export const CarouselContextProvider: FC<
  PropsWithChildren<CarouselContextProviderProps>
> = ({ autoPlay: initialAutoPlay, autoPlayDuration = 3000, children }) => {
  const [activeIndex, setActiveIndex] = useState(0);
  const [autoPlay, setAutoPlay] = useState(initialAutoPlay || false);

  const itemsRef = useRef<(string | number)[]>([]);

  const showNext = () => {
    setActiveIndex((prev) => (prev + 1) % itemsRef.current.length);
  };

  const showPrev = () => {
    setActiveIndex(
      (prev) => (prev - 1 + itemsRef.current.length) % itemsRef.current.length
    );
  };

  const registerItem = (id: string | number) => {
    if (itemsRef.current.includes(id)) {
      return;
    }

    itemsRef.current.push(id);
  };

  useEffect(() => {
    if (!autoPlay) {
      return;
    }

    const interval = setInterval(() => {
      showNext();
    }, autoPlayDuration);

    return () => clearInterval(interval);
  }, [autoPlay, autoPlayDuration]);

  const active = {
    index: activeIndex,
    id: itemsRef.current[activeIndex],
  };

  return (
    <CarouselContext.Provider
      value={{
        autoPlay,
        setAutoPlay,
        active,
        showNext,
        showPrev,
        registerItem,
      }}
    >
      {children}
    </CarouselContext.Provider>
  );
};

export function useCarousel() {
  return useContext(CarouselContext);
}
