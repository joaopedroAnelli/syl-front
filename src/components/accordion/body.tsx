'use client';

import { FC, HTMLAttributes, PropsWithChildren, useContext } from 'react';
import { AccordionContext } from './context';

export type BodyProps = PropsWithChildren<{}>;

export const Body: FC<BodyProps> = ({ children }) => {
  const { open } = useContext(AccordionContext);

  return (
    <div
      data-testid='accordion-body'
      className={`bg-deep-blue transition-all ease-in-out overflow-hidden ${
        open ? 'max-h-96 py-6 md:py-12' : 'invisible max-h-0 p-0'
      } `}
    >
      {children}
    </div>
  );
};
