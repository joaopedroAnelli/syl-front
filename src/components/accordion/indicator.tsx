'use client';

import { useContext, FC, PropsWithChildren } from 'react';
import { AccordionContext } from './context';
import { ChevronDownIcon } from '@heroicons/react/24/solid';

export type IndicatorProps = PropsWithChildren<{}>;

export const Indicator: FC<IndicatorProps> = ({ children }) => {
  const { open } = useContext(AccordionContext);

  return (
    <ChevronDownIcon
      className={`text-off-white h-5 w-5 transition-transform transform ${
        open ? 'rotate-180' : ''
      }`}
    />
  );
};
