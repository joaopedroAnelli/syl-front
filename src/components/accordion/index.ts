import { Root } from './root';
import { Header } from './header';
import { Title } from './title';
import { Indicator } from './indicator';
import { Body } from './body';
import { Description } from './description';

const Accordion = {
  Root,
  Header,
  Title,
  Indicator,
  Body,
  Description,
};

export default Accordion;
