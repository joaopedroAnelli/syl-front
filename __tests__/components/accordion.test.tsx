import '@testing-library/jest-dom';
import { render, screen, act } from '@testing-library/react';
import Accordion from '@/components/accordion';
import { useContext } from 'react';
import { AccordionContext } from '@/components/accordion/context';

describe('Accordion', () => {
  const loremIpsum =
    'Lorem ipsum dolor sit amet consectetur adipisicing elit. Quisquam, voluptate.';

  const AccordionComponent = () => {
    return (
      <Accordion.Root>
        <Accordion.Header>
          <Accordion.Title>Accordion</Accordion.Title>
          <Accordion.Indicator />
        </Accordion.Header>

        <Accordion.Body>
          <Accordion.Description>{loremIpsum}</Accordion.Description>
        </Accordion.Body>
      </Accordion.Root>
    );
  };

  describe('testing functionality', () => {
    it('renders without crashing', () => {
      render(<AccordionComponent />);

      expect(screen.getByText('Accordion')).toBeVisible();

      expect(screen.queryByTestId('accordion-body')).toHaveClass('invisible');
    });

    it('renders with accordion open', () => {
      act(() => {
        render(<AccordionComponent />);
      });

      expect(screen.getByText('Accordion')).toBeVisible();

      expect(screen.queryByTestId('accordion-body')).toHaveClass('invisible');

      act(() => {
        screen.getByRole('button').click();
      });

      expect(screen.getByText(loremIpsum)).toBeVisible();
    });

    it('should close when clicking on the trigger and it is open', () => {
      act(() => {
        render(<AccordionComponent />);
      });

      expect(screen.getByText('Accordion')).toBeVisible();

      expect(screen.queryByTestId('accordion-body')).toHaveClass('invisible');

      act(() => {
        screen.getByRole('button').click();
      });

      expect(screen.getByText(loremIpsum)).toBeVisible();

      act(() => {
        screen.getByRole('button').click();
      });

      expect(screen.queryByTestId('accordion-body')).toHaveClass('invisible');
    });
  });

  describe('testing look', () => {
    it('should be good looking', () => {
      render(<AccordionComponent />);

      const container = screen.getByTestId('accordion-container');

      expect(container).toHaveClass('border-b border-slate-500 bg-deep-blue');

      const title = screen.getByText('Accordion');

      expect(title).toHaveClass('text-off-white text-left');

      const indicator = screen.getByTestId('accordion-indicator');

      expect(indicator).toHaveClass('text-off-white');
    });

    it('should be good looking when open', () => {
      act(() => {
        render(<AccordionComponent />);
      });

      act(() => {
        const trigger = screen.getByRole('button');
        trigger.click();
      });

      const body = screen.getByTestId('accordion-body');

      expect(body).toHaveClass(
        'bg-deep-blue transition-all ease-in-out overflow-hidden max-h-96 py-6 md:py-12'
      );

      const description = screen.getByText(loremIpsum);

      expect(description).toHaveClass('text-slate-500');
    });
  });

  it('context should work fine outside a wrapper', () => {
    const FakeComponent = () => {
      const { open, setOpen } = useContext(AccordionContext);

      setOpen(true);

      return <div>{open ? 'open' : 'closed'}</div>;
    };

    render(<FakeComponent />);
  });

  it('should match snapshot', () => {
    const { container } = render(<AccordionComponent />);

    expect(container).toMatchSnapshot();
  });
});
