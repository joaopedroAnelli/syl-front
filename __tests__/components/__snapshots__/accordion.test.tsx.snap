// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should match snapshot 1`] = `
<div>
  <div
    class="border-b border-slate-500 bg-deep-blue"
    data-testid="accordion-container"
  >
    <button
      class="w-full flex h-16 items-center justify-between"
    >
      <span
        class="text-off-white text-left"
      >
        Accordion
      </span>
      <svg
        aria-hidden="true"
        class="text-off-white h-5 w-5 transition-transform transform "
        data-slot="icon"
        data-testid="accordion-indicator"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M12.53 16.28a.75.75 0 0 1-1.06 0l-7.5-7.5a.75.75 0 0 1 1.06-1.06L12 14.69l6.97-6.97a.75.75 0 1 1 1.06 1.06l-7.5 7.5Z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <div
      class="bg-deep-blue transition-all ease-in-out overflow-hidden invisible max-h-0 p-0 "
      data-testid="accordion-body"
    >
      <p
        class="text-slate-500"
      >
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Quisquam, voluptate.
      </p>
    </div>
  </div>
</div>
`;
